### ISSUE TITLE
How to turn on fzf help completion at zsh?

### ISSUE THREAD

**Issue by Shuraken007** [junegunn/fzf#3349]
Hello, greped a lot in Issues and manual, but can't find how to enable zsh completion for fzf help.
I mean tab complete for fzf options like `--preview`, `--multi`, e.t.c.

shell/completion.bash has function _fzf_opts_completion with opts
![image](https://github.com/junegunn/fzf/assets/23379636/4176a8f4-757e-4dd7-861f-89dafb6db53c)

I havn't find something like this for shell/completion.zsh
![image](https://github.com/junegunn/fzf/assets/23379636/45196886-b7dc-41f6-95b4-c2eabe927b0f)
probably you added some scripts to build help auto?

There are some examples
bat: https://github.com/sharkdp/bat/blob/master/assets/completions/bat.zsh.in
exa: https://github.com/ogham/exa/blob/master/completions/zsh/_exa
ripgrep: https://github.com/BurntSushi/ripgrep/blob/master/complete/_rg

**Comment by LangLangBart** [comment]
No, `fzf` does not have that. Someone has to write a **zsh** completion function for it. 

> _Completion functions for commands are stored in files with names beginning with an underscore , and these files should be placed in a directory listed in the $fpath variable._
> _Source: [HOWTO Guide zsh-users/zsh-completions](https://github.com/zsh-users/zsh-completions/blob/master/zsh-completions-howto.org)_

Some tools have it, for example `gh` or `bat`. If you have `Homebrew` installed, you can see such files or checkout the [zsh-users/zsh-completions](https://github.com/zsh-users/zsh-completions/tree/master/src) github repo.

```sh
ls -1 $(brew --prefix)/share/zsh/site-functions
# _bat
# _brew
# _brew_services
# _cargo
# _delta
# _exa
# _fd
# _gh
# _git
# _hx
# _ninja
# _parallel
# _revolver
# _rg
# _yt-dlp
# _zoxide
```

---

Alternatively, you can run `compdef _gnu_generic fzf` which will search the `fzf` help page and list the flags found, but as the name implies, it is generic and will not work on specific options passed to flags.

```sh
# you may have to run those commands as well
autoload -Uz compinit
compinit

# run
# add the line to your '.zshrc' file if it is useful for your case
compdef _gnu_generic fzf
```

### demo
<img src="https://github.com/junegunn/fzf/assets/92653266/0988cf3f-0ee3-4f63-9cd7-b9913c99df32" width="800">

#### related
- [zsh Manual - Completion System](https://zsh.sourceforge.io/Doc/Release/Completion-System.html#Completion-System)

**Comment by LangLangBart** [comment]
- most zsh completion files for `fzf` [found on GitHub](https://github.com/search?q=%2F%5E%23compdef+fzf%24%2F&type=code) are a bit outdated or just do the same thing as `compdef _gnu_generic fzf` does, but there is **one** version that even makes argument suggestions for certain flags
  - [github.com/lmburns/dotfiles/.config/zsh/completions/_fzf](https://github.com/lmburns/dotfiles/blob/36651e0045db4bfe9c40dd9b1f92028cfcea58b6/.config/zsh/completions/_fzf)

---

- either clone the entire directory and add the path to the `completions` directory to your `$FPATH` or


```sh
git clone --depth 1 https://github.com/lmburns/dotfiles.git lmburns_zsh_completion

# .zshrc
export FPATH="$FPATH:$HOME/lmburns_zsh_completion/.config/zsh/completions"
```


- download this individual file, save it in some folder and add this folder to your `$FPATH`

```sh
cd $HOME/my_completion_folder
curl --remote-name https://raw.githubusercontent.com/lmburns/dotfiles/master/.config/zsh/completions/_fzf

# .zshrc
export FPATH="$FPATH:$HOME/my_completion_folder"
```

**Comment by Shuraken007** [comment]
@LangLangBart - how have you founded this example? Used git search engine for files with name _fzf?

**Comment by LangLangBart** [comment]
### GitHub WebUI
I posted the search link in my previous comment.
See [docs.github.com/understanding-github-code-search-syntax](https://docs.github.com/en/search-github/github-code-search/understanding-github-code-search-syntax) for more details.

---

### Command line

```sh
# uses 'gh', 'fzf' and 'bat'
gh search code $'#compdef fzf' \
	--filename '_fzf' \
	--json repository,path \
	--jq '.[] | [.repository.nameWithOwner, .path] | @tsv' |
	fzf --delimiter '\t' --with-nth 1 \
		--preview 'gh api --cache 10m https://api.github.com/repos/{1}/contents/{2} \
			--jq .content | base64 --decode | bat --language zsh --number' \
		--preview-window 'nohidden:right:wrap:75%' \
		--bind 'ctrl-b:execute-silent:gh browse --repo {1} {2}'
```


<img src="https://github.com/junegunn/fzf/assets/92653266/2c250964-0d77-4828-91e3-d4d72c069aa2" width="800">

**Comment by Shuraken007** [comment]
@LangLangBart  whoa, amazing tricks with that utils. Thanks for this cool staff.
You are on the crest of a wave, cli added `code search` 2 weeks ago.
https://github.com/cli/cli/releases/tag/v2.31.0
had to manually install release from github

I had to use a bit other params for bat `bat -l sh --style=numbers --color=always`

wanna check your dotfiles to get some usefull tricks, but looks like they are private.

**Comment by LangLangBart** [comment]
> _I had to use a bit other params for bat bat --language zsh --style=numbers --color=always_

I missed the `--color=always`, it was only set in my `.zshrc`, and `--number` is just an alias for `--style=numbers`

<img src="https://github.com/junegunn/fzf/assets/92653266/9492e0e7-4089-4c2f-bc87-c81412e72afb" width="600">

> _wanna check your dotfiles to get some usefull tricks, but looks like they are private._

Yes. I use [zsh4humans](https://github.com/romkatv/zsh4humans) as my setup. Checkout the creator's [dotfiles](https://github.com/romkatv/dotfiles-public) - a true `zsh` wizard.

---

PS: The code snippet from my previous comment was transformed into an extension - [gh find-code](https://github.com/LangLangBart/gh-find-code), utilizing the newest features of `fzf`.

<img src="https://github.com/junegunn/fzf/assets/92653266/aacd809e-79d5-451c-93cb-b76160972894" width="600">

**Comment by chrisgrieser** [comment]
I second that having completions would be nice. Even incomplete completions would be better than none at all. Maybe @lmburns would be willing to PR their completions?

@LangLangBart Thanks for various tricks you shared in this thread. `compdef _gnu_generic` and the github code search really useful.

**Comment by iloveitaly** [comment]
https://github.com/RobSis/zsh-completion-generator is another option here for generating completions for tools which don't provide them.

**Comment by chancez** [comment]
Coming at this another angle: could we get the bash completions moved into a separate file https://github.com/junegunn/fzf/blob/cfc37caabcf97f43595b5a06d60d6c7d36ba43e9/shell/completion.bash#L117 and we could reuse that with the zsh bashcompinit?
